server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -354,86 +429,231 @@ public DateFieldMapper.DateFieldType getFieldType() {
+            byte[][] maxs = new byte[ranges.length][];
+            for (int i = 0; i < ranges.length; i++) {
+                double rangeMin = ranges[i].getFrom();
+                double rangeMax = ranges[i].getTo();
+                byte[] min = new byte[byteLen];
+                byte[] max = new byte[byteLen];
+                switch (pointType) {
+                    case "half_float":
+                        HalfFloatPoint.encodeDimension((float) rangeMin, min, 0);
+                        HalfFloatPoint.encodeDimension((float) rangeMax, max, 0);
+                        break;
+                    case "float":
+                        FloatPoint.encodeDimension((float) rangeMin, min, 0);
+                        FloatPoint.encodeDimension((float) rangeMax, max, 0);
+                        break;
+                    case "double":
+                        DoublePoint.encodeDimension(rangeMin, min, 0);
+                        DoublePoint.encodeDimension(rangeMax, max, 0);
+                        break;
+                    case "int":
+                        IntPoint.encodeDimension((int) rangeMin, min, 0);
+                        IntPoint.encodeDimension((int) rangeMax, max, 0);
+                        break;
+                    case "long":
+                        if (this.fieldTypeEnum.isScaled()) {
+                            double scalingFactor = fieldType.getScalingFactor();
+                            rangeMin = scalingFactor * rangeMin;
+                            rangeMax = scalingFactor * rangeMax;
+                        }
+                        LongPoint.encodeDimension((long) rangeMin, min, 0);
+                        LongPoint.encodeDimension((long) rangeMax, max, 0);
+                        break;
+                    case "big_integer":
+                        BigIntegerPoint.encodeDimension(convertDoubleToBigInteger(rangeMin), min, 0);
+                        BigIntegerPoint.encodeDimension(convertDoubleToBigInteger(rangeMax), max, 0);
+                        break;
+                }
 
-        PointValues values = ctx.reader().getPointValues(fastFilterContext.fieldName);
-        if (values == null) return false;
-        // only proceed if every document corresponds to exactly one point
-        if (values.getDocCount() != values.size()) return false;
+                mins[i] = min;
+                maxs[i] = max;
+            }
 
-        NumericDocValues docCountValues = DocValues.getNumeric(ctx.reader(), DocCountFieldMapper.NAME);
-        if (docCountValues.nextDoc() != NO_MORE_DOCS) {
-            logger.debug(
-                "Shard {} segment {} has at least one document with _doc_count field, skip fast filter optimization",
-                fastFilterContext.context.indexShard().shardId(),
-                ctx.ord
-            );
-            return false;
+            return new Ranges(mins, maxs, byteLen);
+        }
+
+        private enum FieldTypeEnum {
```

Comment:
This feels like a lot of logic that we really shouldn't need in this enum.

If you skip support for `scaled_float` (just fall back to the slow logic), you could add a method to `NumberType`, like `byte[] encodePoint(Number number)`. Each specific `NumberType` should be responsible for encoding itself.

Then this enum and much of the logic in `buildRanges` could be replaced with:

```
if (fieldType instanceof NumberType) {
  NumberType numberType = (NumberType) fieldType;
  min = numberType.encodePoint(rangeMin);
  max = numberType.encodePoint(rangeMin);
}
```

Based on the commit message from when `scaled_float` was added, it looks like it was intended as a possible replacement for `half_float`. It looks like it was added to a module to mark it as "experimental", so people wouldn't depend on it. If we're keeping it around, we should move it into `NumberType`.

Reply:
Good idea. Accommodate this by exposing the needed methods from MappedFieldType now.